vault/ssorolecredentialsprovider.go

```diff
@@ -0,0 +1,272 @@
+	StartURL   string
+}
+
+func (p *SSOOIDCProvider) GetAccessToken() (*SSOAccessToken, error) {
+	var (
+		creds = &struct {
+			Token  *SSOAccessToken
+			Client *SSOClientCredentials
+		}{
+			Client: &SSOClientCredentials{},
+			Token:  &SSOAccessToken{},
+		}
+		credsUpdated bool
+	)
+
+	item, err := p.Keyring.Keyring.Get(p.StartURL)
+	if err != nil && err != keyring.ErrKeyNotFound {
+		return nil, err
+	}
+
+	if item.Data != nil {
+		if err = json.Unmarshal(item.Data, &creds); err != nil {
+			return nil, fmt.Errorf("Invalid data in keyring: %v", err)
+		}
+	}
+
+	if creds.Client.Expiration.Before(time.Now()) {
+		creds.Client, err = p.registerNewClient()
+		if err != nil {
+			return nil, err
+		}
+		log.Printf("Created new SSO client for %s (expires at: %s)", p.StartURL, creds.Client.Expiration.String())
+		credsUpdated = true
+	}
+
+	if creds.Token.Expiration.Before(time.Now()) {
+		creds.Token, err = p.createClientToken(creds.Client)
+		if err != nil {
+			return nil, err
+		}
+		log.Printf("Created new SSO access token for %s (expires at: %s)", p.StartURL, creds.Token.Expiration.String())
+		credsUpdated = true
+	}
+
+	if credsUpdated {
+		bytes, err := json.Marshal(creds)
+		if err != nil {
+			return nil, err
+		}
+		err = p.Keyring.Keyring.Set(keyring.Item{
+			Key:                         p.StartURL,
+			Label:                       fmt.Sprintf("aws-vault (%s)", p.StartURL),
+			Data:                        bytes,
+			KeychainNotTrustApplication: true,
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return creds.Token, nil
+}
+
+func (p *SSOOIDCProvider) registerNewClient() (*SSOClientCredentials, error) {
+	c, err := p.OIDCClient.RegisterClient(&ssooidc.RegisterClientInput{
+		ClientName: aws.String(ssoClientName),
+		ClientType: aws.String(ssoClientType),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &SSOClientCredentials{
+		ID:         aws.StringValue(c.ClientId),
+		Secret:     aws.StringValue(c.ClientSecret),
+		Expiration: time.Unix(aws.Int64Value(c.ClientSecretExpiresAt), 0),
+	}, nil
+}
+
+func (p *SSOOIDCProvider) createClientToken(creds *SSOClientCredentials) (*SSOAccessToken, error) {
+	auth, err := p.OIDCClient.StartDeviceAuthorization(&ssooidc.StartDeviceAuthorizationInput{
+		ClientId:     aws.String(creds.ID),
+		ClientSecret: aws.String(creds.Secret),
+		StartUrl:     aws.String(p.StartURL),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Printf(authorizationTemplate, aws.StringValue(auth.VerificationUriComplete))
+	if err := open.Run(aws.StringValue(auth.VerificationUriComplete)); err != nil {
+		log.Printf("failed to open browser: %s", err)
+	}
+
+	var (
+		// These are the default values defined in the following RFC:
+		// https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-3.5
+		slowDownDelay = 5 * time.Second
+		retryInterval = 5 * time.Second
+	)
+	if i := aws.Int64Value(auth.Interval); i > 0 {
+		retryInterval = time.Duration(i) * time.Second
+	}
+
+	for {
```

Comment:
Would it be possible to capture the retry logic in its own separate function? There's an example of this in https://github.com/99designs/aws-vault/blob/877981b40a4614bf5cd921ed63eac96a436b4c97/cli/rotate.go#L149